x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/function/SqlFunctionRegistry.java

```diff
@@ -274,11 +296,147 @@ public SqlFunctionRegistry() {
+    }
+
+    protected interface UnaryDistinctAwareBuilder<T> {
+        T build(Source source, Expression target, Boolean distinct);
+    }
+
+    /**
+     * Build a {@linkplain FunctionDefinition} for a unary function that requires a timezone.
+     */
+    @SuppressWarnings("overloads")  // These are ambiguous if you aren't using ctor references but we always do
+    protected static <T extends Function> FunctionDefinition def(Class<T> function, UnaryZoneIdAwareBuilder<T> ctorRef, String... names) {
+        SqlFunctionBuilder builder = (source, children, cfg, distinct) -> {
+            if (children.size() != 1) {
+                throw new QlIllegalArgumentException("expects exactly one argument");
+            }
+            forbidDistinct(source, distinct);
+            return ctorRef.build(source, children.get(0), cfg.zoneId());
+        };
+        return def(function, builder, true, names);
+    }
+
+    protected interface UnaryZoneIdAwareBuilder<T> {
+        T build(Source source, Expression exp, ZoneId zi);
+    }
+
+    /**
+     * Build a {@linkplain FunctionDefinition} for a binary function that requires a timezone.
+     */
+    @SuppressWarnings("overloads") // These are ambiguous if you aren't using ctor references but we always do
+    protected static <T extends Function> FunctionDefinition def(Class<T> function, BinaryZoneIdAwareBuilder<T> ctorRef, String... names) {
+        SqlFunctionBuilder builder = (source, children, cfg, distinct) -> {
+            if (children.size() != 2) {
+                throw new QlIllegalArgumentException("expects exactly two arguments");
+            }
+            forbidDistinct(source, distinct);
+            return ctorRef.build(source, children.get(0), children.get(1), cfg.zoneId());
+        };
+        return def(function, builder, true, names);
+    }
+
+    protected interface BinaryZoneIdAwareBuilder<T> {
+        T build(Source source, Expression left, Expression right, ZoneId zi);
+    }
+
+    /**
+     * Build a {@linkplain FunctionDefinition} for a three-args function that requires a timezone.
+     */
+    @SuppressWarnings("overloads") // These are ambiguous if you aren't using ctor references but we always do
+    protected static <T extends Function> FunctionDefinition def(Class<T> function,
+                                                                 TertiaryZoneIdAwareBuilder<T> ctorRef,
+                                                                 String... names) {
+        SqlFunctionBuilder builder = (source, children, cfg, distinct) -> {
+            if (children.size() != 3) {
+                throw new QlIllegalArgumentException("expects three arguments");
+            }
+            forbidDistinct(source, distinct);
+            return ctorRef.build(source, children.get(0), children.get(1), children.get(2), cfg.zoneId());
+        };
+        return def(function, builder, true, names);
+    }
+
+    protected interface TertiaryZoneIdAwareBuilder<T> {
```

Comment:
For a three-argument function here you use `tertiary` naming, while in `FunctionRegistry` you use `ternary`. For an uniform approach, could you use the same style in both places? If so, given the rest of the names - unary, binary -, I'd say it should be `ternary`.